### ISSUE TITLE
ZipFile no longer sets the "x" permission when extracting an archive

### ISSUE THREAD

**Issue by eerhardt** [dotnet/corefx#15516]
When running 'dotnet restore' using a CLI that runs on .NET Core 2.0, the execute permission is no longer getting set on the host executable. This is a blocking issue because it means anyone trying to run a self-contained app using the 2.0 CLI will not be able to execute their apps without chmod'ing the published executable.  This is a regression from 1.0.

Steps to repro:

1. Download the following into your project directory and rename it back to ".nupkg".
[runtime.ubuntu.14.04-x64.microsoft.netcore.dotnethost.1.0.1.nupkg.zip](https://github.com/dotnet/corefx/files/733074/runtime.ubuntu.14.04-x64.microsoft.netcore.dotnethost.1.0.1.nupkg.zip)

2. Run the following against netcoreapp1.0 and against netcoreapp2.0

```c#
using System;
using System.IO.Compression;

class Program
{
    static void Main(string[] args)
    {
        ZipFile.ExtractToDirectory(
            "./runtime.ubuntu.14.04-x64.microsoft.netcore.dotnethost.1.0.1.nupkg",
#if NETCOREAPP2_0
            "./NETCOREAPP2_0"
#else
            "./NETCOREAPP1_0"
#endif
		);
    }
}
```

3. Diff the permissions of the extracted `runtimes/ubuntu.14.04-x64/native/dotnet` executable.

Expected result:
The permissions should be identical.

Actual result:

```
Erics-MBP:sandbox eerhardt$ ls -al NETCOREAPP1_0/runtimes/ubuntu.14.04-x64/native/dotnet 
-rwxr--r--  1 eerhardt  staff  128926 Jun 16  2016 NETCOREAPP1_0/runtimes/ubuntu.14.04-x64/native/dotnet
Erics-MBP:sandbox eerhardt$ ls -al NETCOREAPP2_0/runtimes/ubuntu.14.04-x64/native/dotnet 
-rw-r--r--  1 eerhardt  staff  128926 Jun 16  2016 NETCOREAPP2_0/runtimes/ubuntu.14.04-x64/native/dotnet
```

**Comment by stephentoub** [comment]
https://github.com/dotnet/corefx/issues/13342
https://github.com/dotnet/coreclr/pull/7995

**Comment by stephentoub** [comment]
Also:
https://github.com/dotnet/corefx/issues/10223
https://github.com/dotnet/corefx/issues/14853

**Comment by ianhays** [comment]
This is a result of dotnet/coreclr#7995 changing the default filestream file permissions which we use in ZipFile/ZipArchive when extracting.

We aren't going to want to modify the default permissions specifically for ZipArchive, so working around it in CLI seems preferable. Alternatively we could implement either https://github.com/dotnet/corefx/issues/10223 or https://github.com/dotnet/corefx/issues/14853 as Stephen mentioned to get real unix permissions support into ZipArchive. That's something we're going to want to do eventually regardless, but it isn't on the 2.0.0 list yet. Then there's the question of whether we just want to go all out with https://github.com/dotnet/corefx/issues/9225 or https://github.com/dotnet/corefx/issues/15289 and expose some `chmod` equivalent.

**Comment by ianhays** [comment]
>Alternatively we could implement either #10223 or #14853 as Stephen mentioned to get real unix permissions support into ZipArchive. 

Permissions support in ZipArchive won't happen for 2.0.0, so our other options are:

- Modify the permissions in Nuget: https://github.com/NuGet/Home/issues/4424
- Modify them in the CLI around the nuget restore call
- Change the default File permissions for ZipArchive

Seems like the NuGet fix is the only one that makes sense.

@eerhardt @stephentoub any thoughts?

**Comment by ianhays** [comment]
@eerhardt are we good to close this now that https://github.com/dotnet/corefx/pull/18565 is in? I know it's not exactly what you were asking for since [ZipFile automatically applying and recording permissions had to be pushed off into its own issue that won't make it for 2.0.0](https://github.com/dotnet/corefx/issues/17342), but manual permissions recording is at least unblocked.

**Comment by eerhardt** [comment]
Technically, as you've explained it to me, this issue is "by design".  We are intentionally breaking this behavior from .NET Core 1.x to 2.0.

Yes, it is find to close this.  Thank you for getting at least this amount of functionality done for 2.0.

**Comment by ianhays** [comment]
>Technically, as you've explained it to me, this issue is "by design". We are intentionally breaking this behavior from .NET Core 1.x to 2.0.

That is correct, yes. It is an intended result of the change to the default FileStream permissions in 2.0.

>Yes, it is find to close this. Thank you for getting at least this amount of functionality done for 2.0.

I'm glad we were able to figure out something that works and get it merged in time :)

**Comment by karelz** [comment]
@ianhays can you please log the difference here? https://github.com/terrajobst/platform-compat/issues/21

**Comment by tmds** [comment]
@karelz @eerhardt @ianhays I think the permission change is causing two distinct issues:
- https://github.com/dotnet/cli/issues/6438 (dup of this)
- https://github.com/dotnet/cli/issues/6437

I don't think either of those are fixed for 2.0, or I don't understand the fix :)

**Comment by ianhays** [comment]
From hopping around the issues linked to those, it would appear to be more than 2 issues. The takeaway here is that this is an issue that requires solutions in several places:

- CoreFX needs to expose the permissions bits in ZipArchive so that zips made in the future can carry the correct permissions (done - this is represented by this issue)
- NuGet needs to check for these permissions when unpacking. It also needs to handle zips that don't have any permissions. It also needs to write the permissions when packing.
- CLI shouldn't need to do anything special if the above three things are done in NuGet.
